transports/dns/src/lib.rs

```diff
@@ -369,17 +365,19 @@ where
                         Ok(out) => return Ok(out),
                         Err(err) => {
                             tracing::debug!("Dial error: {:?}.", err);
-                            if unresolved.is_empty() {
-                                return Err(err);
-                            }
+                            dial_errors.push(err);
                             if dial_attempts == MAX_DIAL_ATTEMPTS {
                                 tracing::debug!(
                                     "Aborting dialing after {} attempts.",
                                     MAX_DIAL_ATTEMPTS
                                 );
-                                return Err(err);
+                                break;
+                            }
+                            if unresolved.is_empty(){
+                                // If there are no further addresses to try—or we've hit the limit—
+                                // break out of the loop.
+                                break;
                             }
```

Comment:
This needs to be moved back up before checking the limit. The log message about the limit should only be printed if there are more unresolved dials.